torch/_prims/executor.py

```diff
@@ -45,34 +45,52 @@ def call_function(self, target, args, kwargs):
                     args = tuple(map(_to_nvfuser_constant, args))
                     target = target.impl_nvfuser
                     args = (fd,) + args
-                    return target(*args, **kwargs)
+                    nv_args = [to_nv(x) for x in args]
+                    nv_kwargs = {k: to_nv(v) for k, v in kwargs.items()}
+                    return target(*nv_args, **nv_kwargs)
 
             def to_nv(arg):
                 if isinstance(arg, torch.Tensor):
                     x = fd.define_tensor(
                         arg.size(), arg.stride(), getnvFuserDtype(arg.dtype)
                     )
-                    fd.add_input(x)
                     return x
+                elif isinstance(arg, Number):
+                    return fd.define_constant(arg)
                 else:
                     return arg
 
+            def add_inputs(arg):
+                x = to_nv(arg)
+                if isinstance(x, TensorView):
+                    fd.add_input(x)
+                return x
+
             # Transforms graph to call nvfuser lowerings
-            nv_args = tree_map(to_nv, args)
-            nv_kwargs = tree_map(to_nv, kwargs)
+            nv_args = tree_map(add_inputs, args)
+            nv_kwargs = tree_map(add_inputs, kwargs)
```

Comment:
I was entertaining the idea of removing the need to call `add_input` all together since if you are defining a tensor or scalar, it is an input.  Therefore, NVFuser would just handle that accounting.

Reply:
Isn't `FusionDefinition.add_input` required to register the input as an input to the fusion? So that `fusion.execute` would know how to map inputs of the `execute` function to the inputs registered with `add_input`.
There are cases when defined tensor or scalar are not inputs to `fusion.execute` like
```py
def add2(fd, a):
    two_constant = fd.define_constant(2.0)
    fd.Ops.add(a, two_constant)
```

Reply:
There are 3 things that get defined separate from ops and those are Tensors, Scalars, and Constants.  Tensors and Scalars can result in fusion inputs but Constants do not.

In the C++ backend, the call for adding an input would not go away but, as a convenience, `define_tensor` and `define_scalar` could be augmented to additionally call the fusion objects `add_input` instead of having separate calls to `define_tensor` and then `add_input` at the `FusionDefinition` level.

I can't think of a situation where we define Tensors or Scalars where they are not inputs since every where else operations define the intermediates.  It's possible I am missing a case, though.

Reply:
Is there a concept of constant tensor in nvFuser?

Here's a case that I think is easy to find in users' scripts that fails with our current setup (tested on this branch):
```
In [1]: import torch

In [2]: from torch.fx.experimental.proxy_tensor import make_fx

In [3]: from torch._prims.executor import execute

In [4]: def func(a):
   ...:     b = torch.ones(a.shape, device='cuda')
   ...:     return torch._prims.add(a, b)
   ...:

In [5]: a = torch.randn(3, 3, device='cuda')

In [6]: gm = make_fx(func)(a)

In [7]: gm.graph.print_tabular()
opcode         name               target             args                      kwargs
-------------  -----------------  -----------------  ------------------------  --------
placeholder    a_1                a_1                ()                        {}
get_attr       _tensor_constant0  _tensor_constant0  ()                        {}
call_function  add                prims.add          (a_1, _tensor_constant0)  {}
output         output             output             (add,)                    {}

In [8]: execute(gm, a, executor="aten")
Out[8]:
tensor([[-0.0409,  1.5461,  0.0905],
        [ 1.4936,  0.8956,  1.2108],
        [ 1.1867,  1.8426,  1.7041]], device='cuda:0')

In [9]: execute(gm, a, executor="nvfuser")
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
File ~\dev\pytorch\torch\_prims\executor.py:79, in execute(gm, executor, *args, **kwargs)
     77         fd.add_output(o)
---> 79 nv_results = fusion.execute(
     80     tuple(arg for arg in args if isinstance(arg, torch.Tensor))
     81 )
     83 results = []

RuntimeError: thread_predicates_.find(tv_inp) != thread_predicates_.end() INTERNAL ASSERT FAILED at "C:\\Users\\Ivan\\dev\\pytorch\\torch\\csrc\\jit\\codegen\\cuda\\lower_thread_predicate.cpp":224, please report a bug to PyTorch. Thread predicate map was not initialized, couldn't find T1_l[ iblockIdx.x36{( ceilDiv(( ceilDiv(( ceilDiv(( i3 * i4 ), 128) ), 1) ), 1) )}, iUS37{1}, iS35{1}, ithreadIdx.x33{128} ]
```

Reply:
There isn't a constant Tensor because NVFuser doesn't allocate memory.  If it did support an operation like `torch.ones()`, it would likely be through an operator similar to `randLike()` instead of via a constant definition.

Reply:
Not sure if it is a good idea to do so, but I added a new commit https://github.com/pytorch/pytorch/pull/78655/commits/df55ea7a608363eb2290fcafda854d864bb72f7b to handle constant tensor by making it a fusion input.